File: phases/13-concurrent-and-parallel/05-condition-variables-and-monitors/code/main.c
/*
 * Condition Variables and Monitors
 * Phase 13 — Concurrent & Parallel Computing
 *
 * Build It, Steps 1 & 2:
 *   1. Broken Busy-Wait — a consumer spins on a flag (100% CPU).
 *   2. Producer-Consumer with pthread_cond — proper bounded buffer
 *      including lost-wakeup bug, broadcast, and spurious-wakeup handling.
 *
 * Compile:  gcc -std=c11 -Wall -Wextra -pedantic -o main main.c -lpthread
 * Usage:    ./main
 *
 * Step 1 runs first (watch CPU with top/htop), then Step 2 demonstrates
 * the CV-based bounded buffer.
 */

#define _POSIX_C_SOURCE 200809L

#include <pthread.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

/* ================================================================== */
/*  Step 1 — Broken Busy-Wait                                         */
/* ================================================================== */
/*
 * A mutex protects shared data, but it cannot express "wait until X is
 * true."  Without a condition variable, threads must resort to spinning
 * — checking a flag in a tight loop, burning 100% of a CPU core.
 *
 * This step demonstrates the problem.  Watch CPU usage while it runs.
 */

static volatile int busy_flag = 0;

static void* busy_waiter(void* arg) {
    (void)arg;
    fprintf(stderr, "[BusyWait] Consumer spinning... (watch CPU)\n");

    /* BAD: spins at 100% CPU until the flag changes.  No yield,
     * no sleep, no signal — just pure waste. */
    while (busy_flag == 0) {
        /* tight loop — pegs a core */
    }

    fprintf(stderr, "[BusyWait] Consumer: flag is set!  "
            "(1 second of 100%% CPU wasted)\n");
    return NULL;
}

static void* busy_setter(void* arg) {
    (void)arg;
    sleep(1);                   /* guarantee the waiter spins for 1 s */
    busy_flag = 1;
    fprintf(stderr, "[BusyWait] Producer: flag set to 1\n");
    return NULL;
}

static void run_busy_wait(void) {
    pthread_t t1, t2;
    pthread_create(&t1, NULL, busy_waiter, NULL);
    pthread_create(&t2, NULL, busy_setter, NULL);
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    busy_flag = 0;              /* reset for cleanliness */
}

/* ================================================================== */
/*  Step 2 — Bounded Buffer with pthread_cond_t                       */
/* ================================================================== */
/*
 * A bounded (fixed-capacity) buffer shared between producers and
 * consumers.  Uses TWO condition variables:
 *
 *   can_consume  — "the buffer is not empty"
 *   can_produce  — "the buffer is not full"
 *
 * Each cv has its own wait queue: consumers wait on can_consume,
 * producers wait on can_produce.  This avoids the thundering-herd
 * problem where a broadcast would wake both groups unnecessarily.
 *
 * Key correctness properties:
 *   - Every wait() is in a while-loop (Mesa semantics + spurious wakeups).
 *   - Signal is called while holding the lock (lost-wakeup avoidance).
 *   - Signal is called AFTER modifying the state (the condition changed).
 */

#define BUF_CAPACITY 8

typedef struct {
    int buf[BUF_CAPACITY];
    size_t head;                /* read index  (dequeue from head) */
    size_t tail;                /* write index (enqueue at tail)   */
    size_t count;               /* number of items currently in buffer */
    pthread_mutex_t lock;
    pthread_cond_t can_consume; /* "buffer not empty" — consumers wait here */
    pthread_cond_t can_produce; /* "buffer not full"  — producers wait here */
} bounded_buffer_t;

static void bb_init(bounded_buffer_t* bb) {
    memset(bb->buf, 0, sizeof(bb->buf));
    bb->head = 0;
    bb->tail = 0;
    bb->count = 0;
    pthread_mutex_init(&bb->lock, NULL);
    pthread_cond_init(&bb->can_consume, NULL);
    pthread_cond_init(&bb->can_produce, NULL);
}

static void bb_destroy(bounded_buffer_t* bb) {
    pthread_mutex_destroy(&bb->lock);
    pthread_cond_destroy(&bb->can_consume);
    pthread_cond_destroy(&bb->can_produce);
}

/*
 * bb_put — insert an item.  Blocks if the buffer is full.
 *
 * The while-loop is essential: even after pthread_cond_wait returns,
 * the condition (count < capacity) may be false because:
 *   1. Spurious wakeup — the OS woke us for no reason.
 *   2. Mesa semantics — another consumer may have been woken first
 *      and filled the buffer again before we re-acquired the lock.
 */
static void bb_put(bounded_buffer_t* bb, int item) {
    pthread_mutex_lock(&bb->lock);
    while (bb->count == BUF_CAPACITY) {
        /* atomically: unlock + sleep.  On wake: re-lock. */
        pthread_cond_wait(&bb->can_produce, &bb->lock);
    }

    /* Critical section: modify shared state. */
    bb->buf[bb->tail] = item;
    bb->tail = (bb->tail + 1) % BUF_CAPACITY;
    bb->count++;

    /* The condition "buffer not empty" just became true.  Signal one
     * consumer.  The signal is queued only if a consumer is waiting;
     * if none are, the next consumer will find the item immediately. */
    pthread_cond_signal(&bb->can_consume);
    pthread_mutex_unlock(&bb->lock);
}

/*
 * bb_get — remove and return an item.  Blocks if the buffer is empty.
 */
static int bb_get(bounded_buffer_t* bb) {
    pthread_mutex_lock(&bb->lock);
    while (bb->count == 0) {
        pthread_cond_wait(&bb->can_consume, &bb->lock);
    }

    int item = bb->buf[bb->head];
    bb->head = (bb->head + 1) % BUF_CAPACITY;
    bb->count--;

    /* The condition "buffer not full" just became true.  Signal one
     * producer that there's room again. */
    pthread_cond_signal(&bb->can_produce);
    pthread_mutex_unlock(&bb->lock);
    return item;
}

/* ---- Test harness: multiple producers + consumers ---- */

typedef struct {
    bounded_buffer_t* bb;
    int id;
    int items;
} worker_arg_t;

static void* producer_thread(void* arg) {
    worker_arg_t* wa = (worker_arg_t*)arg;
    for (int i = 0; i < wa->items; i++) {
        int val = i + wa->id * 10000;   /* unique per producer */
        bb_put(wa->bb, val);
        fprintf(stderr, "[Producer %d] put %d\n", wa->id, val);
        /* Small delay to make the interleaving interesting. */
        nanosleep(&(struct timespec){0, 5000000L}, NULL); /* 5 ms */
    }
    return NULL;
}

static void* consumer_thread(void* arg) {
    worker_arg_t* wa = (worker_arg_t*)arg;
    for (int i = 0; i < wa->items; i++) {
        int val = bb_get(wa->bb);
        fprintf(stderr, "[Consumer %d] got %d\n", wa->id, val);
        nanosleep(&(struct timespec){0, 3000000L}, NULL); /* 3 ms */
    }
    return NULL;
}

/* ---- Lost-Wakeup Demonstration ---- */
/*
 * THE LOST-WAKEUP BUG:
 * If a thread calls pthread_cond_signal BEFORE any thread is waiting
 * on that CV, the signal is a no-op — it has zero effect.  If the
 * consumer then calls pthread_cond_wait, it will sleep forever because
 * the signal that would wake it already happened and was discarded.
 *
 * This is the single most common bug with condition variables.
 *
 * The correct pattern is:
 *   1. Lock the mutex.
 *   2. Modify shared state.
 *   3. Signal the CV.
 *   4. Unlock the mutex.
 *
 * And on the receiver side:
 *   1. Lock the mutex.
 *   2. while (condition false) { wait(&cv, &lock); }
 *   3. Use the shared state.
 *   4. Unlock.
 *
 * The signal must happen AFTER the wait side is already waiting, or
 * the signal is lost.  In a well-structured program the mutex ensures
 * this: the producer can't signal until it holds the lock, and the
 * consumer releases the lock atomically when it calls wait().
 *
 * But if the producer manages to run the entire critical section
 * (lock → modify → signal → unlock) before the consumer ever enters
 * the wait, the signal is lost.
 */

#if 0
/* Uncomment this block, recompile, and run.  The consumer will hang. */
static void* buggy_producer(void* arg) {
    bounded_buffer_t* bb = (bounded_buffer_t*)arg;

    /* Producer runs FIRST — no consumer is waiting yet. */
    pthread_mutex_lock(&bb->lock);
    bb->buf[bb->tail] = 42;
    bb->tail = (bb->tail + 1) % BUF_CAPACITY;
    bb->count = 1;
    pthread_cond_signal(&bb->can_consume);  /* LOST! nobody listening */
    pthread_mutex_unlock(&bb->lock);
    fprintf(stderr, "[BuggyProducer] signaled — but nobody was listening!\n");
    return NULL;
}

static void* buggy_consumer(void* arg) {
    bounded_buffer_t* bb = (bounded_buffer_t*)arg;
    sleep(1);  /* arrive late — the signal already happened */

    pthread_mutex_lock(&bb->lock);
    while (bb->count == 0) {
        /* MAY WAIT FOREVER — the signal was already sent and lost. */
        pthread_cond_wait(&bb->can_consume, &bb->lock);
    }
    int val = bb->buf[bb->head];
    bb->head = (bb->head + 1) % BUF_CAPACITY;
    bb->count--;
    pthread_mutex_unlock(&bb->lock);
    fprintf(stderr, "[BuggyConsumer] got %d\n", val);
    return NULL;
}

static void run_lost_wakeup(void) {
    bounded_buffer_t bb;
    bb_init(&bb);

    pthread_t p, c;
    /* Start producer FIRST so it runs and signals before consumer waits. */
    pthread_create(&p, NULL, buggy_producer, &bb);
    pthread_create(&c, NULL, buggy_consumer, &bb);
    pthread_join(p, NULL);
    pthread_join(c, NULL);   /* ← this join will hang */

    bb_destroy(&bb);
}
#endif

/* ---- Broadcast Demonstration ---- */
/*
 * pthread_cond_broadcast wakes EVERY thread waiting on the CV.
 * Use broadcast when the condition change could satisfy MULTIPLE
 * waiters.  Example: a single "shutdown" flag that should wake
 * all worker threads.
 *
 * In contrast, pthread_cond_signal wakes exactly one waiter (or
 * none if the queue is empty).  Using signal when multiple waiters
 * could proceed is a correctness bug — only one wakes, the rest
 * sleep forever.
 */

static volatile int shutdown_flag = 0;

typedef struct {
    int id;
    pthread_mutex_t* lock;
    pthread_cond_t* shutdown_cv;
} worker_shutdown_arg_t;

static void* worker_with_shutdown(void* arg) {
    worker_shutdown_arg_t* wa = (worker_shutdown_arg_t*)arg;

    pthread_mutex_lock(wa->lock);
    while (!shutdown_flag) {
        pthread_cond_wait(wa->shutdown_cv, wa->lock);
    }
    pthread_mutex_unlock(wa->lock);

    fprintf(stderr, "[Worker %d] received shutdown signal\n", wa->id);
    return NULL;
}

static void run_broadcast_demo(void) {
    fprintf(stderr, "\n--- Broadcast Demo ---\n");

    pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
    pthread_cond_t shutdown_cv = PTHREAD_COND_INITIALIZER;
    pthread_t workers[4];
    worker_shutdown_arg_t args[4];

    /* Start 4 workers, all waiting on the same CV. */
    for (int i = 0; i < 4; i++) {
        args[i] = (worker_shutdown_arg_t){i, &lock, &shutdown_cv};
        pthread_create(&workers[i], NULL, worker_with_shutdown, &args[i]);
    }

    nanosleep(&(struct timespec){0, 50000000L}, NULL); /* 50 ms — let workers start */

    /* Signal shutdown.  pthread_cond_signal(&shutdown_cv) would wake
     * only ONE worker, leaving the other three asleep forever.
     * broadcast wakes all four. */
    pthread_mutex_lock(&lock);
    shutdown_flag = 1;
    pthread_cond_broadcast(&shutdown_cv);
    pthread_mutex_unlock(&lock);

    for (int i = 0; i < 4; i++)
        pthread_join(workers[i], NULL);

    shutdown_flag = 0;
    pthread_mutex_destroy(&lock);
    pthread_cond_destroy(&shutdown_cv);
}

/* ---- Spurious Wakeup Simulation ---- */
/*
 * POSIX explicitly allows pthread_cond_wait to return even though
 * nobody called signal or broadcast.  This is called a SPURIOUS
 * WAKEUP.  The canonical while-loop pattern handles it.
 *
 * To simulate: we can't force the OS to produce a spurious wakeup,
 * but we can show what happens if a thread is woken and the condition
 * is still false.
 */

typedef struct {
    int id;
    bounded_buffer_t* bb;
} spurious_worker_arg_t;

static void* spurious_consumer(void* arg) {
    spurious_worker_arg_t* wa = (spurious_worker_arg_t*)arg;

    pthread_mutex_lock(&wa->bb->lock);
    /* Without the while-loop, this would crash on spurious wakeup: */
    int count = wa->bb->count;
    if (count == 0) {
        pthread_cond_wait(&wa->bb->can_consume, &wa->bb->lock);
    }
    /* BUG: if this was a spurious wakeup, bb->count is still 0,
     * and the following access reads garbage from the buffer. */
    int item = wa->bb->buf[wa->bb->head];
    pthread_mutex_unlock(&wa->bb->lock);

    fprintf(stderr, "[SpuriousDemo Consumer %d] got %d "
            "(may be garbage from empty buffer!)\n", wa->id, item);
    return NULL;
}

static void show_spurious_wakeup_danger(void) {
    fprintf(stderr, "\n--- Spurious Wakeup Danger ---\n");
    fprintf(stderr, "This code uses if() instead of while().  "
            "On spurious wakeup, it reads garbage.\n");

    bounded_buffer_t bb;
    bb_init(&bb);

    /* Start a consumer that uses the WRONG pattern (if, not while). */
    pthread_t c;
    spurious_worker_arg_t ca = {1, &bb};
    pthread_create(&c, NULL, spurious_consumer, &ca);
    pthread_join(c, NULL);

    /* Note: not actually triggering a spurious wakeup here (the OS
     * doesn't guarantee one), but the code is buggy regardless. */
    fprintf(stderr, "The if-pattern is BROKEN — always use while.\n\n");
    bb_destroy(&bb);
}

/* ---- Integrate Step 2 into a clean run ---- */

static void run_bounded_buffer(void) {
    bounded_buffer_t bb;
    bb_init(&bb);

    pthread_t producers[2];
    pthread_t consumers[3];

    worker_arg_t pa[] = {{&bb, 1, 5}, {&bb, 2, 5}};
    worker_arg_t ca[] = {{&bb, 1, 4}, {&bb, 2, 3}, {&bb, 3, 3}};

    /* Launch 2 producers and 3 consumers.  The buffer capacity is 8,
     * tighter than the 10 total items, so producers will block on
     * can_produce and consumers will block on can_consume — both CVs
     * get exercised. */
    for (int i = 0; i < 2; i++)
        pthread_create(&producers[i], NULL, producer_thread, &pa[i]);
    for (int i = 0; i < 3; i++)
        pthread_create(&consumers[i], NULL, consumer_thread, &ca[i]);

    for (int i = 0; i < 2; i++)
        pthread_join(producers[i], NULL);
    for (int i = 0; i < 3; i++)
        pthread_join(consumers[i], NULL);

    bb_destroy(&bb);
    fprintf(stderr, "[BoundedBuffer] All done.  "
            "%d items produced, %d consumed.\n", 10, 10);
}

/* ================================================================== */
/*  Main                                                              */
/* ================================================================== */

int main(void) {
    fprintf(stderr, "=== Condition Variables and Monitors ===\n\n");

    fprintf(stderr, "=== Step 1: Broken Busy-Wait ===\n");
    fprintf(stderr, "(Watch CPU usage with top/htop/Activity Monitor "
            "— one core will spike to 100%%)\n");
    run_busy_wait();
    fprintf(stderr, "Spinning wastes CPU.  A condition variable "
            "lets the thread sleep instead.\n\n");

    fprintf(stderr, "=== Step 2: Bounded Buffer (pthread_cond) ===\n");
    fprintf(stderr, "Two CVs: can_consume (empty→not-empty) "
            "and can_produce (full→not-full).\n");
    run_bounded_buffer();
    fprintf(stderr, "\n");

    run_broadcast_demo();
    fprintf(stderr, "\n");

    show_spurious_wakeup_danger();

    fprintf(stderr, "\n=== Summary ===\n");
    fprintf(stderr, "  Step 1: 100%% CPU spinning — never do this.\n");
    fprintf(stderr, "  Step 2: pthread_cond_wait/signal — threads sleep "
            "until work arrives.\n");
    fprintf(stderr, "  Broadcast: wakes ALL waiters on a CV.\n");
    fprintf(stderr, "  Key rule: ALWAYS use while(), never if(), "
            "around pthread_cond_wait.\n");
    return 0;
}

File: phases/13-concurrent-and-parallel/05-condition-variables-and-monitors/code/main.rs
// Condition Variables and Monitors
// Phase 13 — Concurrent & Parallel Computing
//
// Build It, Steps 3 & 4:
//   3. Monitor Pattern — struct with Mutex + Condvar, safe API.
//      The monitor encapsulates locking so callers cannot forget to signal.
//   4. Multiple Conditions — bounded queue with separate "has data" /
//      "has space" Condvars.  Producers and consumers each wait on their
//      own CV, avoiding the thundering-herd problem.
//
// Run:  rustc main.rs && ./main     (or  cargo run)

use std::collections::VecDeque;
use std::sync::{Arc, Condvar, Mutex};
use std::thread;
use std::time::Duration;

// ==================================================================
//  Step 3 — Monitor Pattern: Unbounded Channel
// ==================================================================
//
// A monitor is a concurrency abstraction that bundles:
//   1. Shared state (protected by a mutex)
//   2. One or more condition variables
//   3. Methods that expose a safe API
//
// The key advantage: callers of send() and recv() cannot forget to lock
// the mutex or signal the CV.  The monitor's internal methods guarantee
// the protocol.
//
// Rust's Condvar::wait consumes the MutexGuard, so you *cannot* access
// the shared data while the thread is asleep — the guard is gone.
// When wait returns, the guard is re-acquired and returned as a new
// value.  This is the "Guard pattern."

pub struct Channel<T> {
    items: Mutex<Vec<T>>,
    ready: Condvar,
}

impl<T> Channel<T> {
    pub fn new() -> Self {
        Channel {
            items: Mutex::new(Vec::new()),
            ready: Condvar::new(),
        }
    }

    /// Send an item — pushes it into the queue and notifies one waiter.
    ///
    /// The signal happens *after* the push, while still holding the lock.
    /// The waker won't run until we release the lock (Mesa semantics).
    pub fn send(&self, msg: T) {
        let mut guard = self.items.lock().unwrap();
        guard.push(msg);
        // NOTE: notify_one is a no-op if nobody is waiting — that is fine.
        // The item is in the queue; the next recv() call will find it.
        self.ready.notify_one();
    }

    /// Receive an item — blocks until one is available.
    ///
    /// Uses the canonical while-loop pattern even though Condvar::wait
    /// tries to handle spurious wakeups.  The while loop covers:
    ///   1. Spurious wakeups — the OS may wake us for no reason.
    ///   2. Mesa semantics — another consumer may have grabbed the item
    ///      between the signal and our re-acquisition of the lock.
    ///
    /// Note how `guard` is rebound: wait() consumes the old guard and
    /// returns a new one with the lock re-acquired.
    pub fn recv(&self) -> T {
        let mut guard = self.items.lock().unwrap();
        while guard.is_empty() {
            // wait() consumes the guard, atomically unlocks + sleeps.
            // On wake it re-acquires the lock and returns a new guard.
            guard = self.ready.wait(guard).unwrap();
        }
        // guard is re-acquired — safe to access the vector.
        guard.remove(0)
    }
}

/// Drop implementation: nothing special needed since Mutex and Condvar
/// both implement Drop.  The monitor is RAII-clean.
impl<T> Drop for Channel<T> {
    fn drop(&mut self) {
        // In a production channel you might want to notify waiters that
        // the channel is closed so they can exit instead of blocking
        // forever.  For this lesson we accept the simple behavior.
    }
}

fn step3_monitor_pattern() {
    println!("=== Step 3: Monitor Pattern (Rust) ===");

    // A single-producer, single-consumer test.
    let chan = Arc::new(Channel::new());
    let mut handles = vec![];

    // Consumer — will block on recv() until the producer sends.
    let c = chan.clone();
    handles.push(thread::spawn(move || {
        println!("[Consumer] waiting for data...");
        let val = c.recv();
        println!("[Consumer] received {}", val);

        let val2 = c.recv();
        println!("[Consumer] received {}", val2);
    }));

    // Producer — sends two items with a delay.
    let c = chan.clone();
    handles.push(thread::spawn(move || {
        thread::sleep(Duration::from_millis(200));
        c.send(42);
        println!("[Producer] sent 42");

        thread::sleep(Duration::from_millis(100));
        c.send(99);
        println!("[Producer] sent 99");
    }));

    for h in handles {
        h.join().unwrap();
    }
    println!("[Monitor Pattern] OK — both items received.\n");
}

// ==================================================================
//  Step 4 — Multiple Conditions: Bounded Queue
// ==================================================================
//
// The bounded queue uses TWO condition variables:
//
//   can_read  — signaled when data becomes available (not empty)
//   can_write — signaled when space becomes available (not full)
//
// Producers wait on can_write; consumers wait on can_read.  Using
// separate CVs means a consumer never accidentally wakes a producer
// and vice versa.  This is more efficient than a single CV that
// requires broadcast (which wakes *everyone*, most of whom must
// immediately go back to sleep).
//
// The monitor pattern means all locking is internal: callers just
// call push() and pop().  They cannot forget to signal.

pub struct BoundedQueue<T> {
    inner: Mutex<Inner<T>>,
    can_read: Condvar,
    can_write: Condvar,
}

struct Inner<T> {
    data: VecDeque<T>,
    capacity: usize,
}

impl<T> BoundedQueue<T> {
    pub fn new(capacity: usize) -> Self {
        assert!(capacity > 0, "BoundedQueue capacity must be at least 1");
        BoundedQueue {
            inner: Mutex::new(Inner {
                data: VecDeque::with_capacity(capacity),
                capacity,
            }),
            can_read: Condvar::new(),
            can_write: Condvar::new(),
        }
    }

    /// Push an item into the queue, blocking if the queue is full.
    ///
    /// The while-loop handles spurious wakeups and Mesa semantics:
    /// even though we waited on can_write, another producer may have
    /// filled the buffer before we re-acquired the lock.
    pub fn push(&self, item: T) {
        let mut guard = self.inner.lock().unwrap();

        while guard.data.len() == guard.capacity {
            // Wait on the "space available" CV.
            guard = self.can_write.wait(guard).unwrap();
        }

        guard.data.push_back(item);

        // Wake ONE consumer that there's data to read.
        // (notify_all would wake ALL consumers — thundering herd)
        self.can_read.notify_one();
    }

    /// Pop an item from the queue, blocking if the queue is empty.
    pub fn pop(&self) -> T {
        let mut guard = self.inner.lock().unwrap();

        while guard.data.is_empty() {
            // Wait on the "data available" CV.
            guard = self.can_read.wait(guard).unwrap();
        }

        let item = guard.data.pop_front().unwrap();

        // Wake ONE producer that there's space to write.
        self.can_write.notify_one();
        item
    }

    /// Non-blocking try-pop.  Returns None if the queue is empty.
    /// Useful when you want to poll without blocking (e.g., in a UI
    /// event loop).
    pub fn try_pop(&self) -> Option<T> {
        let mut guard = self.inner.lock().unwrap();
        let item = guard.data.pop_front();
        if item.is_some() {
            self.can_write.notify_one();
        }
        item
    }

    /// Return the current length without blocking.
    pub fn len(&self) -> usize {
        let guard = self.inner.lock().unwrap();
        guard.data.len()
    }

    /// Return the capacity.
    pub fn capacity(&self) -> usize {
        let guard = self.inner.lock().unwrap();
        guard.capacity
    }

    /// Block until the queue has at least `n` items, then return them
    /// as a batch.  This reduces lock acquisitions for bulk consumers.
    pub fn pop_batch(&self, n: usize) -> Vec<T> {
        let mut guard = self.inner.lock().unwrap();
        // Wait until we have at least n items.
        while guard.data.len() < n {
            guard = self.can_read.wait(guard).unwrap();
        }
        let mut batch: Vec<T> = Vec::with_capacity(n);
        for _ in 0..n {
            batch.push(guard.data.pop_front().unwrap());
        }
        // Signal producers that space freed up.
        self.can_write.notify_one();
        batch
    }
}

fn step4_multiple_conditions() {
    println!("=== Step 4: Multiple Conditions (BoundedQueue) ===");

    let queue = Arc::new(BoundedQueue::new(3)); // tiny capacity → pressure
    let mut handles = vec![];

    // ==============================================================
    //  2 producers, each pushing 5 items  (10 total)
    // ==============================================================
    for id in 0..2 {
        let q = queue.clone();
        handles.push(thread::spawn(move || {
            for i in 0..5 {
                let val = id * 100 + i;
                q.push(val);
                println!("[Producer {}] pushed {}", id, val);
                thread::sleep(Duration::from_millis(10));
            }
            println!("[Producer {}] done", id);
        }));
    }

    // ==============================================================
    //  3 consumers, each popping 4 or 3 items  (10 total)
    // ==============================================================
    for id in 0..3 {
        let q = queue.clone();
        let count = if id == 0 { 4 } else { 3 };
        handles.push(thread::spawn(move || {
            for _ in 0..count {
                let val = q.pop();
                println!("[Consumer {}] popped {}", id, val);
                thread::sleep(Duration::from_millis(15));
            }
            println!("[Consumer {}] done", id);
        }));
    }

    for h in handles {
        h.join().unwrap();
    }
    println!("[BoundedQueue] All 10 items produced and consumed.");

    // ==============================================================
    //  Demonstrate try_pop and pop_batch
    // ==============================================================
    println!("\n--- try_pop and pop_batch demo ---");

    let q = BoundedQueue::new(10);

    // Queue starts empty → try_pop returns None.
    assert!(q.try_pop().is_none());
    println!("try_pop on empty queue: None (correct)");

    // Push 5 items.
    for i in 0..5 {
        q.push(i);
    }
    println!("After pushing 5 items, len = {}", q.len());

    // try_pop now succeeds.
    let val = q.try_pop().unwrap();
    println!("try_pop returned Some({})", val);
    println!("len after try_pop = {}", q.len());

    // pop_batch(3) — blocks until 3 items available (they already are).
    let batch = q.pop_batch(3);
    println!("pop_batch(3) returned: {:?}", batch);
    println!("len after pop_batch = {}", q.len());

    println!("\n[BoundedQueue] All extra demos passed.\n");
}

// ==================================================================
//  Main
// ==================================================================

fn main() {
    println!("=== Condition Variables and Monitors (Rust) ===\n");

    step3_monitor_pattern();
    step4_multiple_conditions();

    println!("=== Summary ===");
    println!("  Step 3: Monitor pattern — safe send/recv API with Condvar.");
    println!("           Guard pattern: wait() consumes + returns MutexGuard.");
    println!("  Step 4: BoundedQueue with separate can_read / can_write Condvars.");
    println!("           try_pop() and pop_batch() for flexible use.");
    println!("  Key rule: ALWAYS loop, even with Rust's type-safe Condvar.");
}
